public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx

```diff
@@ -23,94 +113,173 @@ const LabelsField: FC<Props> = ({ className }) => {
+                      })}
+                      defaultValue={field.value ? { label: field.value, value: field.value } : undefined}
+                      options={values}
+                      onChange={(newValue: SelectableValue) => {
+                        setValue(`labels.${index}.value`, newValue.value);
+                      }}
+                      onOpenMenu={() => {
+                        setSelectedKey(labels[index].key);
+                      }}
+                      type="value"
+                    />
+                  </Field>
+
+                  <RemoveButton className={styles.deleteLabelButton} index={index} remove={remove} />
+                </div>
+              </div>
+            );
+          })}
+          <AddButton className={styles.addLabelButton} append={append} />
+        </>
+      )}
+    </>
+  );
+};
+
+const LabelsWithoutSuggestions: FC = () => {
+  const styles = useStyles(getStyles);
+  const {
+    register,
+    control,
+    watch,
+    formState: { errors },
+  } = useFormContext<RuleFormValues>();
+
+  const labels = watch('labels');
+  const { fields, remove, append } = useFieldArray({ control, name: 'labels' });
+
+  return (
+    <>
+      {fields.map((field, index) => {
+        return (
+          <div key={field.id}>
+            <div className={cx(styles.flexRow, styles.centerAlignRow)}>
+              <Field
+                className={styles.labelInput}
+                invalid={!!errors.labels?.[index]?.key?.message}
+                error={errors.labels?.[index]?.key?.message}
+              >
+                <Input
+                  {...register(`labels.${index}.key`, {
+                    required: { value: !!labels[index]?.value, message: 'Required.' },
+                  })}
+                  placeholder="key"
+                  data-testid={`label-key-${index}`}
+                  defaultValue={field.key}
+                />
+              </Field>
+              <InlineLabel className={styles.equalSign}>=</InlineLabel>
+              <Field
+                className={styles.labelInput}
+                invalid={!!errors.labels?.[index]?.value?.message}
+                error={errors.labels?.[index]?.value?.message}
+              >
+                <Input
+                  {...register(`labels.${index}.value`, {
+                    required: { value: !!labels[index]?.key, message: 'Required.' },
+                  })}
+                  placeholder="value"
+                  data-testid={`label-value-${index}`}
+                  defaultValue={field.value}
+                />
+              </Field>
+              <RemoveButton className={styles.deleteLabelButton} index={index} remove={remove} />
+            </div>
+          </div>
+        );
+      })}
+      <AddButton className={styles.addLabelButton} append={append} />
+    </>
+  );
+};
+
+const LabelsField: FC<Props> = ({ className, suggest, dataSourceName }) => {
```

Comment:
I wonder if we should just have two separate components `LabelsWithSuggestions` and `LabelsWithoutSuggestions` and remove `LabelsField` wdyt? The one with suggestions does quite a lot of things under the hood compared to the one without suggestions so maybe we could just keep them separate

Reply:
I thought about this and I agreed with you, especially since I didn't like having both the `suggest` and `dataSourceName` props as they were too tied to each other. 
As I was about to implement it I thought of something else though: have the `dataSourceName` prop be optional in the `LabelsField` component and depending on its value/lack of, show either the dropdowns with suggestions or the text inputs if absent. This allowed me to remove the `suggest` prop. This way, we have the `LabelsField` component  in control of what to show based on whether it receives a `dataSourceName` or not. I kind of like having this logic encapsulated there.
What do you think? Let me know if you disagree as I have no problem changing it to two separate components though.